### ISSUE TITLE
SIGABRT when calling emsdk/upstream/bin/wasm-opt:  uncaught exception of type NonconstantException

### ISSUE THREAD

**Issue by michaelfominwahoo** [emscripten-core/emscripten#17594]
I recently upgraded emsdk from `3.1.3` to `3.1.18` and now see a new build error.

```
libc++abi: terminating due to uncaught exception of type wasm::ConstantExpressionRunner<wasm::PrecomputingExpressionRunner>::NonconstantException
emcc: error: '/Users/misha/myapp/submodules/emsdk/upstream/bin/wasm-opt --strip-dwarf --post-emscripten -O3 --low-memory-unused --zero-filled-memory --strip-debug --strip-producers /Users/misha/myapp/out/CruxTrackModule.wasm -o /Users/misha/myapp/out/CruxTrackModule.wasm --mvp-features' failed (received SIGABRT (-6))
```

The error was first introduced for me in `3.1.9` as I verified all releases from `3.1.3` to `3.1.8` built fine for me

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.18 (d5ca9bba6513763d5bdddbd0efff759332bd85d7)
clang version 16.0.0 (https://github.com/llvm/llvm-project 48129cf0ed5053ed3fdf4f928180635e84892614)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /Users/misha/myapp/submodules/emsdk/upstream/bin
```

**Failing command line in full:**
The project is really huge, so I'm apprehensive about adding the entire output here.  Hopefully, this helps kick things off.
```
emcc \
  -o "${out_js_path}" \
  ${HEADER_LIST} \
  ${SRC_LIST} \
  -O3 \
  -s WASM=1 \
  -s MODULARIZE=1 \
  -s EXPORT_NAME="${MODULE_NAME}" \
  -s ASSERTIONS=1 \
  -s EXPORT_ES6=1 \
  -s USE_ES6_IMPORT_META=0 \
  -s NO_EXIT_RUNTIME=1 \
  -s ALLOW_TABLE_GROWTH \
  -s ALLOW_MEMORY_GROWTH=1 \
  -s EXPORTED_RUNTIME_METHODS=[${EMSCRIPTEN_FUNCTIONS_STR_COMMA%,}] \
  -s EXPORTED_FUNCTIONS=[${FUNCTIONS_LIST%,}]
```

**Full link command and output with `-v` appended:**
The project is really huge, so I'm apprehensive about adding the entire output here.  Hopefully, the above helps.

**Comment by sbc100** [comment]
Could you attach just the input file to wasm-opt (`/Users/misha/myapp/out/CruxTrackModule.wasm`) so something can reproduce that crash?

**Comment by michaelfominwahoo** [comment]
Sure thing.

[CruxTrackModule.wasm.zip](https://github.com/emscripten-core/emscripten/files/9278493/CruxTrackModule.wasm.zip)

**Comment by sbc100** [comment]
Fancy taking a look at a binaryen crash @tlively @kripken ?

**Comment by kripken** [comment]
Running `wasm-opt --strip-dwarf --post-emscripten -O3 --low-memory-unused --zero-filled-memory --strip-debug --strip-producers --mvp-features CruxTrackModule.wasm` does not crash for me. It also runs ok in valgrind.

Does running the individual `wasm-opt` like that crash for you @michaelfominwahoo ? If so then since this is a regression, I'd recommend bisecting on Binaryen to find where this broke.

**Comment by sbc100** [comment]
@kripken did you try specifically with wasm-opt from the 3.1.18 SDK?

**Comment by kripken** [comment]
Yes, and also latest `main`. Neither crashes for me.

I tested on Linux.

**Comment by michaelfominwahoo** [comment]
I'm running on Mac if that helps.
I ran the command (from the output) directly from the terminal and got the same result
```
libc++abi: terminating due to uncaught exception of type wasm::ConstantExpressionRunner<wasm::PrecomputingExpressionRunner>::NonconstantException
zsh: abort      /Users/misha/myapp/systm/crux_tools/submodules/emsdk/upstream/bin/wasm-opt  
```

**Comment by kripken** [comment]
Ok, it could be a mac-specific regression then. I don't have a mac, so the best thing would probably be to bisect. @michaelfominwahoo you can do that either on Binaryen itself, or [using the emsdk](https://emscripten.org/docs/contributing/developers_guide.html#bisecting) (the second way you don't need to do any building from source yourself).

**Comment by sbc100** [comment]
Is this the M1 mac issue maybe?  (the one that got hotfixed in 3.1.18?) Can you try reinstalling 3.1.18 or installing `tot`?

**Comment by michaelfominwahoo** [comment]
Progress, the issue is happening on my new M1 Mac, and not happening on my old Intel-based Mac.
I re-installed 3.1.18 and got the same error...
I installed tot, using `emsdk install tot`, and got a crazy different error.
[tot.crash.txt](https://github.com/emscripten-core/emscripten/files/9290447/tot.crash.txt)
Sorry, I have not had the time today to bisect, which is kinda new to me. I'm guessing it's the way to find the exact commit things broke.

**Comment by jinh0** [comment]
I'm having the same issue, installed emscripten 3.1.18 for the first time, and received a similar error message as @michaelfominwahoo :

```
libc++abi: terminating due to uncaught exception of type wasm::ConstantExpressionRunner<wasm::PrecomputingExpressionRunner>::NonconstantException
emcc: error: '/Users/jinhoyoon/dev/verovio/emscripten/emsdk/upstream/bin/wasm-opt --strip-dwarf --post-emscripten -O3 --low-memory-unused --zero-filled-memory --strip-debug --strip-producers build/verovio.wasm -o build/verovio.wasm --mvp-features' failed (received SIGABRT (-6))
 Failed.
```

**Comment by michaelfominwahoo** [comment]
3.1.17 working for me too! 
Sorry, when I upgraded from 3 > 18, and had problems, I tried 4, 5, 6, 7, 8 (all ok), then 9 failed and I assumed the issue was introduced in 9... I did not think to try 17.

**Comment by ephemer** [comment]
FWIW I'm getting the same error on 3.1.18 (original M1 MacBook Pro).

I would try 3.1.17 but I get the following error if I try:

```
% bazel build --config wasm -c opt hello-world
INFO: Repository emscripten_bin_mac_arm64 instantiated at:
  /Users/geordiejay/dev/tflite-bazel/WORKSPACE:9:16: in <toplevel>
  /Users/geordiejay/dev/tflite-bazel/emscripten_deps.bzl:138:21: in emscripten_deps
Repository rule http_archive defined at:
  /private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/bazel_tools/tools/build_defs/repo/http.bzl:353:31: in <toplevel>
WARNING: Download from https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/d27fef2070c86a218965da8b8b5df8b4425aa3bb/wasm-binaries-arm64.tbz2 failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was 40d16dcf284286a18ee0b1de55db9dc33ebd8d6e181bf8e32a1fd2adb0ab55f3 but wanted 5f60d3f351d06d862e853a294642d24243d6cb197e34c2f2602d80555c2eb014
ERROR: An error occurred during the fetch of repository 'emscripten_bin_mac_arm64':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/bazel_tools/tools/build_defs/repo/http.bzl", line 100, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/d27fef2070c86a218965da8b8b5df8b4425aa3bb/wasm-binaries-arm64.tbz2] to /private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/emscripten_bin_mac_arm64/temp11992825463725974979/wasm-binaries-arm64.tbz2.tar.bz2: Checksum was 40d16dcf284286a18ee0b1de55db9dc33ebd8d6e181bf8e32a1fd2adb0ab55f3 but wanted 5f60d3f351d06d862e853a294642d24243d6cb197e34c2f2602d80555c2eb014
ERROR: /Users/geordiejay/dev/tflite-bazel/WORKSPACE:9:16: fetching http_archive rule //external:emscripten_bin_mac_arm64: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/bazel_tools/tools/build_defs/repo/http.bzl", line 100, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/d27fef2070c86a218965da8b8b5df8b4425aa3bb/wasm-binaries-arm64.tbz2] to /private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/emscripten_bin_mac_arm64/temp11992825463725974979/wasm-binaries-arm64.tbz2.tar.bz2: Checksum was 40d16dcf284286a18ee0b1de55db9dc33ebd8d6e181bf8e32a1fd2adb0ab55f3 but wanted 5f60d3f351d06d862e853a294642d24243d6cb197e34c2f2602d80555c2eb014
ERROR: /Users/geordiejay/dev/tflite-bazel/BUILD:37:6: //:compiler_files depends on @emscripten_bin_mac_arm64//:compiler_files in repository @emscripten_bin_mac_arm64 which failed to fetch. no such package '@emscripten_bin_mac_arm64//': java.io.IOException: Error downloading [https://storage.googleapis.com/webassembly/emscripten-releases-builds/mac/d27fef2070c86a218965da8b8b5df8b4425aa3bb/wasm-binaries-arm64.tbz2] to /private/var/tmp/_bazel_geordiejay/f08b4496e6845be1e5ab4320687adc85/external/emscripten_bin_mac_arm64/temp11992825463725974979/wasm-binaries-arm64.tbz2.tar.bz2: Checksum was 40d16dcf284286a18ee0b1de55db9dc33ebd8d6e181bf8e32a1fd2adb0ab55f3 but wanted 5f60d3f351d06d862e853a294642d24243d6cb197e34c2f2602d80555c2eb014
ERROR: Analysis of target '//hello-world:hello-world' failed; build aborted:
INFO: Elapsed time: 9,588s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (25 packages loaded, 97 targets configured)
```

**Comment by dmitru** [comment]
Same error on 3.1.18 on M1 mac, gonna try 3.1.17 to see if it helps.

**Comment by sbc100** [comment]
@dschuff did the 3.1.18 hotfix for M1 Mac not work maybe?

**Comment by dschuff** [comment]
I just double-checked and it seems to work for me. If you are using emsdk from git, make sure you are at revision 4a48a752e6a8bef6f222622f2b4926d5eb3bdeb3 or later (currently that's the HEAD of main)

**Comment by michaelfominwahoo** [comment]
Yeah that is the revision I'm at, tagged 3.1.18. It's the latest on main.
`4a48a752e6a8bef6f222622f2b4926d5eb3bdeb3`

**Comment by sbc100** [comment]
Can you try with `emsdk install tot`?

**Comment by dschuff** [comment]
I just pushed out the 3.1.19 release, and I tested this example on my M1 with 3.1.19, which works. I'm going to assume that the issue you saw was the same as the known issue with 3.1.18, and is fixed in 3.1.19, and mark this issue as fixed. Please reopen it if you have the same issue with 3.1.19.

**Comment by gogins** [comment]
I have tried `emsdk install tot` and things are now building with emsdk on MacBook Pro M1.
